packages/realm-react/src/useQuery.tsx

```diff
@@ -29,25 +39,33 @@ import { symbols } from "@realm/common";
  */
 export function createUseQuery(useRealm: () => Realm) {
   return function useQuery<T>(
-    type: string | ({ new (...args: any): T } & Realm.ObjectClass),
-  ): Realm.Results<T & Realm.Object> {
+    ...[type, collectionCallback, dependencies]: UseQueryArgs<T>
+  ): RealmResults<T> {
     const realm = useRealm();
 
+    const collectionCallbackMemo = useMemo(() => {
+      if (!!collectionCallback) {
+        return collectionCallback;
+      }
+      return (collection: RealmResults<T>) => collection;
+    }, dependencies || []);
```

Comment:
```suggestion
        const queryResult = useMemo(() => {
          return query(realm.objects(type));
        }, deps)
```
With the defaults, this memo becomes simpler.  I also think we can call the callback directly and use the result directly in the lines below.
Also with the change above, deps include the `query` so that if that callback changes, the results are recalculated.

Reply:
If you are going to use realm inside this useMemo then `realm` and `type` will need to be dependencies along side `deps`

I think this complicates the use of the dependency array here and if the goal is to have a memoized `queryResult` it should be in a separate `useMemo` that uses the queryCallback as a dependency like this

```js
    const queryCallback = useMemo(() => {
      // query is unstable so we memoize it here
      return query;
    }, deps);

    const queryResult = useMemo(() => {
      return queryCallback(realm.objects(type));
    }, [type, realm, queryCallback]);
    
    ...
    
    const { collection, tearDown } = useMemo(() => {
      return createCachedCollection({
        collection: queryResult,
        realm,
        updateCallback: forceRerender,
        updatedRef,
      });
    }, [queryResult]);
```